File: AcWing 785. 快速排序.cpp
//https://www.acwing.com/activity/content/problem/content/819/
#include <iostream>
using namespace std;

#define N 100010

int n;
int a[N];

void quick_sort(int* a, int l, int r)
{   
    if(l == r)
        return;
    int i = l-1, j = r+1, k = a[l + r >> 1];
    while(i < j)
    {
        do i++; while(a[i] < k);
        do j--; while(a[j] > k);
        if(i < j) swap(a[i], a[j]);
    }
    quick_sort(a, l, j), quick_sort(a, j+1, r);
}

int main()
{
    cin >> n;
    for(int i = 0 ;i < n; i++)
    {
        cin >> a[i];
    }
    quick_sort(a, 0, n-1);
    for(int i = 0 ;i < n; i++)
    {
        cout << a[i] << " ";
    }
}

File: AcWing 785. 快速排序.py
# https://www.acwing.com/activity/content/problem/content/819/
def quick_sort(a, l, r):
    if l == r:
        return
    i = l-1
    j = r+1
    k = a[l + r >> 1]
    while i < j:
        i += 1
        while a[i] < k:
            i += 1
        j -= 1
        while a[j] > k:
            j -= 1
        if i < j:
            temp = a[i]
            a[i] = a[j]
            a[j] = temp
    quick_sort(a, l, j)
    quick_sort(a, j+1, r)
            
n = int(input())
a = []
s = input()
s = s.strip()
l = s.split(" ")
for i in l:
    a.append(int(i))
quick_sort(a, 0, n-1)
for i in a:
    print(i, end=" ")